FlowCal/gate.py

```diff
@@ -334,11 +352,19 @@ def density2d(data, channels=[0,1],
     if data_ch.shape[0] <= 1:
         raise ValueError('data should have more than one event')
 
-    # Extract default bins if necessary
-    if (bins is None and hasattr(data_ch, 'hist_bin_edges')
-            and data_ch.hist_bin_edges(0) is not None
-            and data_ch.hist_bin_edges(1) is not None):
-        bins = np.array(data_ch.hist_bin_edges())
+    # If bins are not specified, try to get bins from data object
+    if (hasattr(data_ch, 'hist_bins')
+            and data_ch.hist_bins(0) is not None
+            and data_ch.hist_bins(1) is not None):
+        # Duplicate bins into a two-element list if necessary
+        if not isinstance(bins, list):
+            bins = [bins, bins]
+        # X Axis
+        if bins[0] is None or isinstance(bins[0], int):
+            bins[0] = data_ch.hist_bins(0, nbins=bins[0], log=xlog)
+        # Y axis
+        if bins[1] is None or isinstance(bins[1], int):
+            bins[1] = data_ch.hist_bins(1, nbins=bins[1], log=ylog)
```

Comment:
same comments as below for `plot.density2d()`
>* i don't understand this section. you were originally checking if `bins` is `None` (i.e. only extracting bins from data object if `bins` was not specified). Shouldn't this still be the desired behavior?
* can you add the name of the first argument to `hist_bins()`? (as as you've done with the other 2 arguments) i forgot what that was and had to go look it up.
* was unclear what `data_plot.hist_bins(#)` was doing. still not 100% sure. why are you checking if that's none? when does that function return `None`?
* description of `bins` also didn't make sense. Seems like you should only extract `bins` from data when `bins` is `None`
* does `isinstance(bins,list)` fail with tuples or array-like? should that just be checking for iterable?